talk/basicconcepts/control.tex

```diff
@@ -49,23 +49,24 @@
     \end{cppcode*}
     \vspace{-0.2cm}
     \begin{itemize}
-      \item if test is \mintinline{cpp}{true} expression1 is returned
-      \item else expression2 is returned
+      \item If test is \mintinline{cpp}{true} expression1 is returned
+      \item Else expression2 is returned
     \end{itemize}
   \end{block}
   \pause
   \begin{exampleblock}{Practical example}
     \begin{cppcode*}{}
-      int collatz(int a) {
-        return a==1 ? 1 : collatz(a%2==0 ? a/2 : 3*a+1);
-      }
+     unsigned int collatz(unsigned int a) {
+      return a == 1 ? 1 : collatz(a & 1 ? 3*a + 1 : a/2);
+     }
```

Comment:
I am not sure this version is better. Compilers will generate the bit op for you from the modulo op. I would argue the modulo is easier to read as. I guess the `unsigned` was chosen because of the bit op.

Reply:
Most compilers optimize both versions to `return 1`, funnily enough. However, when they don't, my version has slightly shorter assembly. https://godbolt.org/z/5c46cnYYf In any case, I just replaced `a%2==0` with `a & 1` so that I could put spaces to format the code a bit better. Since bitwise operators are discussed nearby in the contents, I thought this would be an ok tradeoff to make for readability. The unsigned was chosen because a check for `a < 0` was removed compared with the previous slide. Using unsigned makes that impossible (this is in the commit message).

Reply:
That optimization is indeed surprising :D If you compare just the disassembly for the implementation of an `is_even` (https://godbolt.org/z/968nEeqsq), which is what the condition of the ternary is, the compiler generates the same code.

The main point for me is to not encourage using a bitwise op over a modulo when testing whether an integral is even. Could you afford a linebreak to keep the better spacing and leave the `a % 2 == 0`?

Reply:
I don't see a problem with the bitwise operator, modulo is really slow if one is not careful. I think I will just drop this commit from my pull request.

Reply:
You are absolutely right! I am just talking about the case to express whether an integral is even, where I prefer the modulo for readability. Maybe others want to share their opinion? @hageboeck ?

Reply:
In terms of readability:
Don't use collatz at all. Something *much simpler* might illustrate the usage better than something that you have to parse recursively in your head.

Reply:
e.g.
```c++
std::cout << "The condition is " << (condition ? "true" : "false")
```

Reply:
I agree, but this is out of scope for my changes, I will just drop the commit so we can proceed to merge the other changes.